Transit filters for GTFS API's `planConnection`

leonardehrenfried opened the pull request:
### Summary

This adds the filter API originally developed for the Transmodel API to the GTFS API.

It allows you to included or exclude routes or agencies from the routing result.

For example it allows you to express the following filter criteria:

- give me only results for agency A1
- give me only results for routes R1, R2
- exclude results from agency A2, A3
- exclude results from routes A4, A5 and agencies A4
- ...

### Issue

#6239

As discussed in this issue, the schema docs was copied more or less from the Transmodel API and uses the same semantics. Some features like dated service journeys and group of lines, which don't have a GTFS equivalent, are omitted.

### Unit tests

Added.

### Documentation

Javadoc and schema docs.

t2gran commented on application/src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -4291,6 +4296,22 @@ input PlanDateTimeInput @oneOf {
+"A collection of selectors for what routes/agencies should be included in or excluded from the search."
+input PlanFilterInput {
+  "A list of selectors for what routes/agencies should be excluded during the search. If route/agency/mode matches with at least one selector it will be excluded. An empty list or a list containing `null` is forbidden."
+  not: [PlanFilterSelectInput!]
+  "A list of selectors for what routes/agencies should be allowed during search. In order to be accepted a route/agency has to match with at least one selector. An empty list or a list containing `null` is forbidden."
+  select: [PlanFilterSelectInput!]
```
Consider using "exclude/include" instead of "not/select" - I do not have an opinion, and I have not check if there are semantic differences.

t2gran commented on utils/src/test/java/org/opentripplanner/utils/collection/CollectionUtilsTest.java:
```diff
@@ -65,4 +73,14 @@ void testIsEmptyCollection() {
+    CollectionUtils.requireNullOrNonEmpty(null, "test");
+    CollectionUtils.requireNullOrNonEmpty(List.of(1), "test");
+
+    assertThrows(IllegalArgumentException.class, () -> {
+      CollectionUtils.requireNullOrNonEmpty(List.of(), "test");
+    });
```
```suggestion
    var ex = assertThrows(IllegalArgumentException.class, () -> {
      CollectionUtils.requireNullOrNonEmpty(List.of(), "test");
    });
    assertEquals("...", ex.getMessage());
```

t2gran commented:
Note! I am not reviewing this, just commented on a few things I noticed while reading through part of the code - core domain only.

leonardehrenfried commented:
@fpurcell this might also be relevant for you.

optionsome commented on application/src/test/java/org/opentripplanner/apis/gtfs/GraphQLFormattingTest.java:
```diff
@@ -7,12 +7,13 @@
-  public static final File SCHEMA_FILE = new File(
-    "src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls"
-  );
+  public static final File SCHEMA_FILE = ResourceLoader.of(
+    GraphQLFormattingTest.class
+  ).mainResourceFile("schema.graphqls");
```
If I remember correctly, there was some reason why I didn't use the resource loader here (maybe because this depends on the built resources). Were these types of issues somehow fixed with this mainResourceFile method? Although, it could be that I just copied the approach from the transmodel API without paying any attention.

leonardehrenfried replied:
I think the method didn't exist back then.

optionsome reviewed (commented):
Was it possible to define these filters in the router configuration? If it is, we probably should expose the defaults in the schema.

leonardehrenfried commented:
While moving the filters into the transit preferences I notices a problem with how they interact with mode settings. I reworked the mapping, added tests and improved the docs.

This is now ready for review.

optionsome commented on application/src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -4603,6 +4603,42 @@ input TransferPreferencesInput {
   slack: Duration
 }
 
+"A collection of selectors for what routes/agencies should be included in or excluded from the search."
+input TransitFilterInput {
+  """
+  A list of selectors for what routes/agencies should be allowed during search.
+    
+  In order to be excluded a route/agency has to match with at least one selector.
+    
+  An empty list or a list containing `null` is forbidden.
+  """
+  exclude: [TransitFilterSelectInput!]
+  """
+  A list of selectors for what routes/agencies should be excluded during the search.
+    
+  If route/agency matches with at least one selector it will be included.
+    
+  An empty list or a list containing `null` is forbidden.
+  """
+  include: [TransitFilterSelectInput!]
+}
```
We should document the priority of include/exclude. If something is included but also excluded, the exclude is the one that rules.

optionsome commented on application/src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -4603,6 +4603,42 @@ input TransferPreferencesInput {
+
+"""
+A list of selectors for filter including or excluding entities from the routing results. Null
+means everything is allowed to be returned and empty lists a not allowed.
```
```suggestion
means everything is allowed to be returned and empty lists are not allowed.
```

optionsome commented on application/src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -4603,6 +4603,42 @@ input TransferPreferencesInput {
+A list of selectors for filter including or excluding entities from the routing results. Null
+means everything is allowed to be returned and empty lists a not allowed.
+
+An entity will match with selectors if it matches with all non-empty lists.
```
When we just have agencies and routes to select, this model is a bit clumsy. I think we could use a `@oneOf` here until we include something where combining multiple fields makes sense. I don't think removing the `@oneOf` restriction is a breaking change (but I'm not 100% sure).

leonardehrenfried replied:
I think it helps people to make a logic mistake so I've added it.

I, too, believe it would not be a breaking change to remove it, so lets add it.

optionsome commented on application/src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -4603,6 +4603,42 @@ input TransferPreferencesInput {
+
+An entity will match with selectors if it matches with all non-empty lists.
+
+The `include` is always applied first, then `exclude`. If only `exclude` is present, the exclude is
+applied to the existing set of results.
```
Feels like this should be on the upper type, not on this. Also, to me this wording is a bit vague. I think the "first" could be misinterpreted as a priority.

optionsome commented on application/src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -4603,6 +4603,42 @@ input TransferPreferencesInput {
+applied to the existing set of results.
+"""
+input TransitFilterSelectInput {
+  "Set of ids for agencies that should be included in/excluded from search"
+  agencies: [String!]
+  "Set of ids for routes that should be included in/excluded from search"
+  routes: [String!]
```
Usually we include an example of what the id is (for example "i.e. value of field `gtfsId` (ID format is `FeedId:StopId`)")

optionsome commented on application/src/test/java/org/opentripplanner/apis/gtfs/GraphQLUtilsTest.java:
```diff
@@ -0,0 +1,16 @@
+package org.opentripplanner.apis.gtfs;
+
+import static org.junit.jupiter.api.Assertions.*;
```
We usually avoid wildcard imports.

optionsome commented on application/src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -4603,6 +4603,54 @@ input TransferPreferencesInput {
+A list of selectors for including or excluding entities from the routing results. Null
+means everything is allowed to be returned and empty lists are not allowed.
+
+An entity will match with selectors if it matches with all lists.
```
What lists do you refer to here? On this level, it's not possible to define multiple lists anymore (for now) due to the `@oneOf` directive.

optionsome commented on application/src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -4618,6 +4666,50 @@ input TransitPreferencesInput {
   """
   board: BoardPreferencesInput
+  """
+  A list of filters for which trips should be included or excluded. A trip will be included if it matches 
+  with at least one filter.
```
If we have one "filter" which defines an include that includes a trip, but also another that defines an exclude. Then I guess it's not enough that the trip matches the include filter?

optionsome replied:
Or maybe this is correct, otherwise it wouldn't be possible to include only route from an agency? So for example:
```
  filters: [
    {
      include: {
        routes: ["F:route1"]
      }
    },
    { 
      exclude: {
        agencies: ["F:agency1"]
      }
    }
  }]
  ```

optionsome commented on application/src/main/resources/org/opentripplanner/apis/gtfs/schema.graphqls:
```diff
@@ -4618,6 +4666,50 @@ input TransitPreferencesInput {
+    
+  Several filters can be defined and form an OR-condition.
+    
+  The following pseudo-GraphQL example means "include all trips with `route1` _or_ `agency1`":
```
This isn't pseudo-GraphQL anymore, it at least seems to be proper GraphQL.

optionsome reviewed (approved):
This feels ok, but I'm still a bit uns
